src/frontend/metrics/AppDiagnosticMetrics.ts

```diff
@@ -0,0 +1,204 @@
+  if (!lastKnownPosition) return undefined;
+  const {latitude, longitude} = lastKnownPosition.coords;
+  const countries = positionToCountries(latitude, longitude);
+  return first(countries);
+}
+
+async function generateAppDiagnosticMetricsData(): Promise<AppDiagnosticMetricsReport> {
+  const result: AppDiagnosticMetricsReport = {
+    dateGenerated: formatIsoUtc(new Date()),
+    os: Platform.OS,
+    osVersion: Platform.Version,
+    deviceLocale: getDeviceLanguageTag(),
+    appLocale: usePersistedLocale.getState().locale,
+    country: await getCountry(),
+  };
+
+  setIfNotNull(result, 'appId', Application.applicationId);
+  setIfNotNull(result, 'appName', Application.applicationName);
+  setIfNotNull(
+    result,
+    'nativeApplicationVersion',
+    Application.nativeApplicationVersion,
+  );
+  setIfNotNull(result, 'nativeBuildVersion', Application.nativeBuildVersion);
+
+  return result;
+}
+
+function saveQueueToStorage(queue: Readonly<AppDiagnosticMetricsQueue>): void {
+  storage.set(STORAGE_KEY, JSON.stringify(queue));
+}
+
+export class AppDiagnosticMetrics {
+  #isEnabled = false;
+  #isOnline = false;
+
+  #subscriptionCleanupFns: null | (() => void)[] = null;
+
+  #update;
+  #updateQueue = new OneAtATimeQueue();
+
+  constructor() {
+    this.#update = () => {
+      this.#updateQueue.add(() => this.#doUpdate());
+    };
+  }
+
+  setEnabled(isEnabled: boolean): void {
+    this.#isEnabled = isEnabled;
+
+    if (isEnabled) {
+      this.#addListenersIfNecessary();
+    } else {
+      this.#removeListeners();
+    }
+
+    this.#update();
+  }
+
+  #addListenersIfNecessary(): void {
+    if (this.#subscriptionCleanupFns) return;
+
+    const subscriptionCleanupFns: (() => void)[] = [];
+
+    subscriptionCleanupFns.push(
+      NetInfo.addEventListener(state => {
+        this.#isOnline = Boolean(state.isInternetReachable);
+        this.#update();
+      }),
+    );
+
+    const appStateSubscription = AppState.addEventListener(
+      'change',
+      this.#update,
+    );
+    subscriptionCleanupFns.push(() => appStateSubscription.remove());
+
+    // We could be smarter about this. For example, we could set the timer for
+    // beginning of the next day. However, timeouts are a bit unreliable (for
+    // example, if the app goes into the background), so we just check
+    // periodically. This also simplifies the code a bit.
+    setInterval(this.#update, CHECK_INTERVAL);
+
+    this.#subscriptionCleanupFns = subscriptionCleanupFns;
+  }
+
+  #removeListeners(): void {
+    const subscriptionCleanupFns = this.#subscriptionCleanupFns ?? [];
+    for (const fn of subscriptionCleanupFns) fn();
+  }
+
+  /**
+   * Update the queue and send metrics as needed.
+   *
+   * Should only be run once at a time. See `this.#update` for a queued alternative.
+   */
+  async #doUpdate(): Promise<void> {
+    if (!this.#isEnabled) {
+      storage.delete(STORAGE_KEY);
+      return;
+    }
+
+    let queue = loadQueueFromStorage();
```

Comment:
This _could_ get slow as reports build up and/or we add things to reports. We could load this once at app initialization, and then only write when it changes. This is a low-priority optimization. I think it is how the persistent storage that we already use works.